### ISSUE TITLE
Accuracy on MNIST is 0

### ISSUE THREAD

**Issue by shilinni** [tensorflow/privacy#48]
I have run mnist_dpsgd_tutorial_keras.py without changing any parameters, except the epoch is set to 1, but the result is like "3000/3000 [==============================] - 153s 51ms/sample - loss: 2.1090 - acc: 0.0000e+00". And its test accuracy is 10%. If I only train it on  digit 0 and digit 1, the training accuracy is around 80%, but if I train on other digits, the training accuracy is 0.

**Comment by schien1729** [comment]
Thank you for the report-- we'll take a look.

**Comment by shilinni** [comment]
Ohhh... It might be something related to my own model. So I have two models:

```
class LocalLinearModel(LocalModel):
    def __init__(self, input_dim=2, output_dim=2,flags=None):
        # Build model
        model = Sequential()
        model.add(Dense(input_dim=input_dim, units=output_dim,
                        activation="softmax", kernel_initializer="normal"))
        ......
        ......
        if flags['dpsgd']:
            dp_average_query = ........
            optimizer =........
            loss = .....
        else:
            loss='categorical_crossentropy'
            optimizer = .......
        model.compile(loss=loss, optimizer=optimizer, metrics=['accuracy'])
        self.model = model

    def fit(self, X, y, epochs=30):
        y_categorical = to_categorical(y, num_classes=self.output_dim)
        self.model.fit(X, y_categorical, epochs=epochs)

    def predict_classes(self, X):
        return self.model.predict_classes(X)
```
The above one has never met this error. But the model below has the problem. 

```
class LocalDenseModel(LocalModel):
    def __init__(self, input_dim=2, output_dim=2,input_shape=(2,),flags=None,optimizer='adam'):
        .....
        model = tf.keras.Sequential([
        tf.keras.layers.Dense(8,input_shape=input_shape),
        tf.keras.layers.Dense(output_dim,activation='softmax')])
        
       ......
        if flags['dpsgd']:
            dp_average_query =.......
            optimizer = .....
            loss = ......
        else:
            optimizer =......
            loss = tf.keras.losses.CategoricalCrossentropy(from_logits=True)
        model.compile(optimizer=optimizer, loss=loss, metrics=['accuracy'])
       
        self.model=model
        
    def fit(self, X, y, epochs=10):
        y_categorical = to_categorical(y, num_classes=self.output_dim)
        self.model.fit(X, y_categorical,epochs=epochs)
        
    def predict_classes(self, X):
        return self.model.predict_classes(X)  
```

So basically the only difference is self.model and loss. I am quite curious at why the second one will have a training accuracy of 0.

**Comment by npapernot** [comment]
In the second model, it looks like you are defining the loss using logits as the input because `from_logits=True`, yet the last layer of the model has a softmax activation (`activation='softmax'`). Could that be creating numerical instabilities that prevent the model from training?

**Comment by shilinni** [comment]
> In the second model, it looks like you are defining the loss using logits as the input because `from_logits=True`, yet the last layer of the model has a softmax activation (`activation='softmax'`). Could that be creating numerical instabilities that prevent the model from training?

Thanks for the reply! I guess this is the problem too. I just don't understand why sometimes the training accuracy is exactly 0 and sometimes it's working fine. It's really weird lol.
